weave-js/src/components/PagePanelComponents/Home/Browse3/typeViews/File/FileView.tsx

```diff
@@ -0,0 +1,251 @@
+  mode?: string;
+  data: FileTypePayload;
+};
+
+type FileMetadata = {
+  original_path: string;
+  mimetype: string;
+  size: number;
+};
+
+export const FileView = ({entity, project, data}: FileViewProps) => {
+  const {useFileContent} = useWFHooks();
+
+  const metadata = useFileContent({
+    entity,
+    project,
+    digest: data.files['metadata.json'],
+  });
+
+  if (metadata.loading) {
+    return <LoadingDots />;
+  }
+  // TODO: Should add an explicit error condition to useFileContent
+  if (metadata.result == null) {
+    return <span>Metadata not found</span>;
+  }
+
+  // Convert ArrayBuffer to JSON
+  let metadataJson;
+  try {
+    const decoder = new TextDecoder();
+    const jsonString = decoder.decode(metadata.result);
+    metadataJson = JSON.parse(jsonString);
+  } catch (error) {
+    console.error('Error parsing metadata JSON:', error);
+    return <span>Error parsing metadata</span>;
+  }
+
+  const file = data.files['file'];
+  return (
+    <FileViewMetadataLoaded
+      entity={entity}
+      project={project}
+      metadata={metadataJson}
+      file={file}
+    />
+  );
+};
+
+type FileViewMetadataLoadedProps = {
+  entity: string;
+  project: string;
+  metadata: FileMetadata;
+  file: string;
+};
+
+const FileViewMetadataLoaded = ({
+  entity,
+  project,
+  metadata,
+  file,
+}: FileViewMetadataLoadedProps) => {
+  const [fileResult, setFileResult] = useState<Blob | null>(null);
+  const [isDownloading, setIsDownloading] = useState(false);
+  const [showPreview, setShowPreview] = useState(false);
+
+  const {useFileContent} = useWFHooks();
+  const {original_path, size, mimetype} = metadata;
+
+  const iconStart = <Icon name={getIconName(mimetype)} />;
+  const filename = original_path.split('/').pop() || original_path;
+  const onDownloadCallback = useCallback((e: React.MouseEvent) => {
+    e.preventDefault();
+    setIsDownloading(true);
+  }, []);
+  const fileContent = useFileContent({
+    entity,
+    project,
+    digest: file,
+    skip: !isDownloading,
+  });
+
+  // Store the last non-null file content result in state
+  // We do this because passing skip: true to useFileContent will
+  // result in fileContent.result getting returned as null even
+  // if it was previously downloaded successfully.
+  useEffect(() => {
+    if (fileContent.result) {
+      const blob = new Blob([fileContent.result], {
+        type: mimetype,
+      });
+
+      setFileResult(blob);
+      setIsDownloading(false);
+    }
+  }, [fileContent.result, mimetype]);
+
+  const doSave = useCallback(() => {
+    if (!fileResult) {
+      console.error('No file result');
+      return;
+    }
+    const url = URL.createObjectURL(fileResult);
+    const a = document.createElement('a');
+    a.href = url;
+    a.download = filename;
+    document.body.appendChild(a);
+    a.click();
+    setTimeout(() => {
+      document.body.removeChild(a);
+      window.URL.revokeObjectURL(url);
+    });
+  }, [fileResult, filename]);
```

Comment:
I don't understand what this does.

Reply:
This method downloads the blob that is in fileResult in a way that works across browsers.

Reply:
I see. Can we extract it from the component as a generic function and make `doSave` call it:
```typescript
  const doSave = useCallback(() => {
    const url = URL.createObjectURL(fileResult);
    downloadBlob(url);
  }, [fileResult, filename]);
```
Then it can be reused elsewhere if needed. Maybe also call `doSave` differently, like `doDownload` or something.

Reply:
I extracted a `saveBlob` function and added a comment about it. I think "Save" is more accurate than "Download" because at this point the data has already been downloaded from the server, and what you are really doing is saving it from memory to the downloads folder.

Longer term I'd recommend extending the server APIs to include one that sets `Content-Disposition: attachment`.